src/js/flows/search/handler.ts

```diff
@@ -19,37 +24,55 @@ export function handle(request: any) {
 
     const flushBufferLazy = whenIdle(flushBuffer)
 
+    let isErrSet = false
     function errored(e) {
       flushBufferLazy.cancel()
       if (abortError(e)) {
+        response.emit("abort")
+        response.emit("status", "SUCCESS")
         resolve()
       } else {
+        isErrSet = true
         response.emit("status", "ERROR")
         response.emit("error", e)
         reject(e)
       }
     }
+
+    const recordCb = createRecordCallback()
+
     request
       .then((stream) => {
         stream
           .callbacks()
-          .start(({task_id}) => {
-            response.emit("start", task_id)
+          .start(() => {
+            response.emit("start")
             response.emit("status", "FETCHING")
           })
-          .records((args: RecordsCallbackArgs) => {
-            channels.set(args.channel, args)
+          .channelSet(({channel_id}) => {
+            currentChannel = channel_id
+          })
+          .record((rec) => {
+            if (currentChannel === null)
+              throw new Error("record received before channel was set")
+            channels.set(currentChannel, recordCb(rec, currentChannel))
             flushBufferLazy()
           })
-          .end(({id, error}) => {
-            if (error) return errored(error)
+          .channelEnd(({channel_id}) => {
             flushBuffer()
-            response.emit("status", "SUCCESS")
-            response.emit("end", id)
-            setTimeout(resolve)
+            response.emit("chan-end", channel_id)
           })
-          .warnings(({warning}) => response.emit("warnings", warning))
+          .warning((pl) => response.emit("warning", pl.warning))
           .error(errored)
+          .internalError(errored)
```

Comment:
This is supposed to be used when there is an internal error in the client, rather than the more explicit "QueryError" type we may get back as a response from the backend. They ended up having the same handler, so maybe that's overkill though.